usr/man/man/usr/share/man/man1/cksum.1
'\" te
.\"  Copyright (c) 1992, X/Open Company Limited All Rights Reserved
.\" Portions Copyright (c) 2009, 2011, Oracle and/or its affiliates. All rights reserved.
.\" Portions Copyright (c) 1982-2007 AT&T Knowledge Ventures
.\" Sun Microsystems, Inc. gratefully acknowledges The Open Group for permission to reproduce portions of its copyrighted documentation. Original documentation from The Open Group can be obtained online at http://www.opengroup.org/bookstore/.
.\" The Institute of Electrical and Electronics Engineers and The Open Group, have given us permission to reprint portions of their documentation. In the following statement, the phrase "this text" refers to portions of the system documentation. Portions of this text are reprinted and reproduced in electronic form in the Sun OS Reference Manual, from IEEE Std 1003.1, 2004 Edition, Standard for Information Technology -- Portable Operating System Interface (POSIX), The Open Group Base Specifications Issue 6, Copyright (C) 2001-2004 by the Institute of Electrical and Electronics Engineers, Inc and The Open Group. In the event of any discrepancy between these versions and the original IEEE and The Open Group Standard, the original IEEE and The Open Group Standard is the referee document. The original Standard can be obtained online at http://www.opengroup.org/unix/online.html.  This notice shall appear on any product containing this material.
.TH cksum 1 "25 Jul 2011" "SunOS 5.11" "User Commands"
.SH NAME
cksum \- write file checksums and sizes
.SH SYNOPSIS
.LP
.nf
\fB/usr/bin/cksum\fR [\fIfile\fR...]
.fi

.SH DESCRIPTION
.sp
.LP
The \fBcksum\fR command calculates and writes to standard output a cyclic redundancy check (CRC) for each input file, and also writes to standard output the number of octets in each file.
.sp
.LP
For each file processed successfully the \fBcksum\fR method writes in the following format:
.sp
.LP
\fB"%u %d %s\en"\fR <\fIchecksum\fR>, <\fI# of octets\fR>, <\fIpath name\fR>
.sp
.LP
If no file operand was specified, the path name and its leading space is omitted.
.sp
.LP
The CRC used is based on the polynomial used for CRC error checking in the referenced Ethernet standard.
.sp
.LP
The encoding for the CRC checksum is defined by the generating polynomial:
.sp
.in +2
.nf
G(x) = x**32 + x**26 + x**23 + x**22+ x**16 + x**12 + x**11
+ x**10 + x**8 + x**7 + x**5 + x**4 + x**2 + x + 1
.fi
.in -2

.sp
.LP
Mathematically, the CRC value corresponding to a given file is defined by the following procedure:
.RS +4
.TP
1.
The \fIn\fR bits to be evaluated are considered to be the coefficients of a mod 2 polynomial \fIM\fR(\fIx\fR) of degree \fIn\fR\(mi\fI1.\fR These \fIn\fR bits are the bits from the file, with the most significant bit being the most significant bit of the first octet of the file and the last bit being the least significant bit of the last octet, padded with zero bits (if necessary) to achieve an integral number of octets, followed by one or more octets representing the length of the file as a binary value, least significant octet first. The smallest number of octets capable of representing this integer is used.
.RE
.RS +4
.TP
2.
\fIM\fR(\fIx\fR) is multiplied by \fIx\fR ^\fI32\fR (that is, shifted left 32 bits) and divided by \fIG\fR(\fIx\fR) using mod 2 division, producing a remainder \fIR\fR(\fIx\fR) of degree \(<= 31.
.RE
.RS +4
.TP
3.
The coefficients of \fIR\fR(\fIx\fR) are considered to be a 32-bit sequence.
.RE
.RS +4
.TP
4.
The bit sequence is complemented and the result is the CRC.
.RE
.SH OPERANDS
.sp
.LP
The following operand is supported:
.sp
.ne 2
.mk
.na
\fB\fIfile\fR\fR
.ad
.RS 8n
.rt  
A path name of a file to be checked. If no \fIfile\fR operands are specified, the standard input is used.
.RE

.SH USAGE
.sp
.LP
The \fBcksum\fR command is typically used to quickly compare a suspect file against a trusted version of the same, such as to ensure that files transmitted over noisy media arrive intact. However, this comparison cannot be considered cryptographically secure. The chances of a damaged file producing the same CRC as the original are astronomically small; deliberate deception is difficult, but probably not impossible.
.sp
.LP
Although input files to \fBcksum\fR can be any type, the results need not be what would be expected on character special device files. Since this document does not specify the block size used when doing input, checksums of character special files need not process all of the data in those files.
.sp
.LP
The algorithm is expressed in terms of a bitstream divided into octets. If a file is transmitted between two systems and undergoes any data transformation (such as moving 8-bit characters into 9-bit bytes or changing \fBLittle Endian\fR byte ordering to \fBBig Endian\fR), identical CRC values cannot be expected. Implementations performing such transformations can extend \fBcksum\fR to handle such situations.
.sp
.LP
See \fBlargefile\fR(5) for the description of the behavior of \fBcksum\fR when encountering files greater than or equal to 2 Gbyte (2^31 bytes).
.SH ENVIRONMENT VARIABLES
.sp
.LP
See \fBenviron\fR(5) for descriptions of the following environment variables that affect the execution of \fBcksum\fR: \fBLANG\fR, \fBLC_ALL\fR, \fBLC_CTYPE\fR, \fBLC_MESSAGES\fR, and \fBNLSPATH\fR.
.SH EXIT STATUS
.sp
.LP
The following exit values are returned:
.sp
.ne 2
.mk
.na
\fB\fB0\fR\fR
.ad
.RS 6n
.rt  
All files were processed successfully.
.RE

.sp
.ne 2
.mk
.na
\fB\fB>0\fR\fR
.ad
.RS 6n
.rt  
An error occurred.
.RE

.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(5) for descriptions of the following attributes:
.sp

.sp
.TS
tab() box;
cw(2.75i) |cw(2.75i) 
lw(2.75i) |lw(2.75i) 
.
ATTRIBUTE TYPEATTRIBUTE VALUE
_
Availabilitysystem/core-os
_
Interface StabilityCommitted
_
StandardSee \fBstandards\fR(5).
.TE

.SH SEE ALSO
.sp
.LP
\fBdigest\fR(1), \fBsum\fR(1), \fBbart\fR(1M), \fBattributes\fR(5), \fBenviron\fR(5), \fBlargefile\fR(5), \fBstandards\fR(5)


usr/src/cmd/cksum/cksum.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright (c) 1994, 2011, Oracle and/or its affiliates. All rights reserved.
 */

#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <locale.h>
#include <libgen.h>

#define	NBUF	4096		/* Read this much at a time */
#define	BSIZE	512		/* Bytes per block */

typedef	unsigned long	CRC;	/* 32 bit (at least) crc size */

static	uchar_t	buf[NBUF];	/* Input buffer */
static	uchar_t	tflag;		/* More than one input file */
static	uchar_t	many;
static	CRC	filecrc;	/* CRC for this file */
static	CRC	totcrc;		/* Total of all files */
static	long long totbytes;	/* Total bytes read */
static	ulong_t	bsize = 1;	/* block size */

static	int	sum(const char *fn);
static	int	usage(void);
static	CRC	docrcposix(CRC crcval, uchar_t *bp, ulong_t n);
static	CRC	postprocess(CRC S, long long n);

static	CRC	(*crcfunc)(CRC crcval, uchar_t *bp, ulong_t n) = docrcposix;
static void	m_crcposix(ulong_t *, const uchar_t *, size_t);

/*
 * Mainline for cksum.
 */
int
main(int argc, char *argv[])
{
	int	c;
	int	exstat = 0;

	(void) setlocale(LC_ALL, "");

#if !defined(TEXT_DOMAIN)
#define	TEXT_DOMAIN	"SYS_TEST"
#endif
	(void) textdomain(TEXT_DOMAIN);

	tflag = many = 0;
	totbytes = totcrc = 0;
	while ((c = getopt(argc, argv, "")) != -1) {
		switch (c) {
		case 't':		/* this is non-functional for now */
			tflag++;
			break;
		default:
			return (usage());
		}
	}

	/* Sum from standard input? */
	if ((many = optind < argc) == 0)
		return (sum(NULL));

	/* Sum each file argument. */
	for (; optind < argc; ++optind)
		exstat |= sum(argv[optind]);

	if (tflag) {
		(void) printf("%lu", postprocess(totcrc, totbytes));
		(void) printf(gettext("\t%lld\t   Total\n"),
		    (long long)((totbytes + bsize - 1) / bsize));
	}
	return (exstat);
}

/*
 * Compute checksum for one file.
 */
static int
sum(const char *fn)
{
	int	n;
	int	fd;
	long long bytes = 0;

	if (fn == NULL || (fn[0] == '-' && fn[1] == '\0'))
		fd = 0;
	else if ((fd = open(fn, O_RDONLY)) < 0) {
		(void) fprintf(stderr, gettext("cksum: input file \"%s\": "),
		    fn);
		perror("");
		return (1);
	}
	filecrc = 0;
	while ((n = read(fd, (char *)buf, sizeof (buf))) > 0) {
		bytes += n;
		filecrc = (*crcfunc) (filecrc, buf, n);
		if (tflag)
			totcrc = (*crcfunc) (totcrc, buf, n);
	}
	(void) close(fd);
	totbytes += bytes;
	(void) printf("%lu", postprocess(filecrc, bytes));
	(void) printf("\t%lld", (long long)((bytes + bsize - 1) / bsize));
	if (many) {
		(void) printf("\t%s", (fn[0] == '-' && fn[1] == '\0') ?
		    gettext("[standard input]") : fn);
	}
	if (n < 0)
		(void) fputs(gettext(" [read error]"), stdout);
	(void) putchar('\n');
	return (n == 0 ? 0 : 1);
}


/*
 * Do CRC-POSIX function by calling a library entry point that has a
 * slightly different calling sequence.
 */
static	CRC
docrcposix(CRC crcval, uchar_t *bp, ulong_t n)
{
	m_crcposix(&crcval, bp, n);
	return (crcval);
}

/*
 * Sum algorithms require various kinds of post-processing.
 * The 'S' and 'R' variables are from the POSIX.2 (Draft 8?) description
 * of the "sum" utility.
 * This code makes no assumption about word size except that a
 * "CRC" type holds at least 32 bits.
 */
static	CRC
postprocess(CRC S, long long n)
{
	/*
	 * POSIX tacks on significant bytes of the length so that
	 * different length sequences of '\0' have different sums;
	 * then it complements sum.
	 */
	uchar_t	char_n[sizeof (n)];
	ulong_t	i;

	for (i = 0; n != 0; n >>= 8, ++i)
		char_n[i] = n & 0xFF;
	return (~docrcposix(S, char_n, i));
}

/*
 * usage message for cksum
 */
static int
usage()
{
	(void) fprintf(stderr, gettext("Usage: cksum [file ...]\n"));
	return (2);
}

/*
 * crcposix -- compute posix.2 compatable 32 bit CRC
 *
 * The POSIX.2 (draft 10) CRC algorithm.
 * This is a 32 bit CRC with polynomial
 *	x**32 + x**26 + x**23 + x**22 + x**16 + x**12 + x**11 + x**10 +
 *	x**8 + x**7 + x**5 + x**4 + x**2 + x**1 + x**0
 */

/* layout is from the POSIX.2 Rationale */
static ulong_t crctab_posix[256] = {
	0x00000000,
	0x04C11DB7, 0x09823B6E, 0x0D4326D9, 0x130476DC, 0x17C56B6B,
	0x1A864DB2, 0x1E475005, 0x2608EDB8, 0x22C9F00F, 0x2F8AD6D6,
	0x2B4BCB61, 0x350C9B64, 0x31CD86D3, 0x3C8EA00A, 0x384FBDBD,
	0x4C11DB70, 0x48D0C6C7, 0x4593E01E, 0x4152FDA9, 0x5F15ADAC,
	0x5BD4B01B, 0x569796C2, 0x52568B75, 0x6A1936C8, 0x6ED82B7F,
	0x639B0DA6, 0x675A1011, 0x791D4014, 0x7DDC5DA3, 0x709F7B7A,
	0x745E66CD, 0x9823B6E0, 0x9CE2AB57, 0x91A18D8E, 0x95609039,
	0x8B27C03C, 0x8FE6DD8B, 0x82A5FB52, 0x8664E6E5, 0xBE2B5B58,
	0xBAEA46EF, 0xB7A96036, 0xB3687D81, 0xAD2F2D84, 0xA9EE3033,
	0xA4AD16EA, 0xA06C0B5D, 0xD4326D90, 0xD0F37027, 0xDDB056FE,
	0xD9714B49, 0xC7361B4C, 0xC3F706FB, 0xCEB42022, 0xCA753D95,
	0xF23A8028, 0xF6FB9D9F, 0xFBB8BB46, 0xFF79A6F1, 0xE13EF6F4,
	0xE5FFEB43, 0xE8BCCD9A, 0xEC7DD02D, 0x34867077, 0x30476DC0,
	0x3D044B19, 0x39C556AE, 0x278206AB, 0x23431B1C, 0x2E003DC5,
	0x2AC12072, 0x128E9DCF, 0x164F8078, 0x1B0CA6A1, 0x1FCDBB16,
	0x018AEB13, 0x054BF6A4, 0x0808D07D, 0x0CC9CDCA, 0x7897AB07,
	0x7C56B6B0, 0x71159069, 0x75D48DDE, 0x6B93DDDB, 0x6F52C06C,
	0x6211E6B5, 0x66D0FB02, 0x5E9F46BF, 0x5A5E5B08, 0x571D7DD1,
	0x53DC6066, 0x4D9B3063, 0x495A2DD4, 0x44190B0D, 0x40D816BA,
	0xACA5C697, 0xA864DB20, 0xA527FDF9, 0xA1E6E04E, 0xBFA1B04B,
	0xBB60ADFC, 0xB6238B25, 0xB2E29692, 0x8AAD2B2F, 0x8E6C3698,
	0x832F1041, 0x87EE0DF6, 0x99A95DF3, 0x9D684044, 0x902B669D,
	0x94EA7B2A, 0xE0B41DE7, 0xE4750050, 0xE9362689, 0xEDF73B3E,
	0xF3B06B3B, 0xF771768C, 0xFA325055, 0xFEF34DE2, 0xC6BCF05F,
	0xC27DEDE8, 0xCF3ECB31, 0xCBFFD686, 0xD5B88683, 0xD1799B34,
	0xDC3ABDED, 0xD8FBA05A, 0x690CE0EE, 0x6DCDFD59, 0x608EDB80,
	0x644FC637, 0x7A089632, 0x7EC98B85, 0x738AAD5C, 0x774BB0EB,
	0x4F040D56, 0x4BC510E1, 0x46863638, 0x42472B8F, 0x5C007B8A,
	0x58C1663D, 0x558240E4, 0x51435D53, 0x251D3B9E, 0x21DC2629,
	0x2C9F00F0, 0x285E1D47, 0x36194D42, 0x32D850F5, 0x3F9B762C,
	0x3B5A6B9B, 0x0315D626, 0x07D4CB91, 0x0A97ED48, 0x0E56F0FF,
	0x1011A0FA, 0x14D0BD4D, 0x19939B94, 0x1D528623, 0xF12F560E,
	0xF5EE4BB9, 0xF8AD6D60, 0xFC6C70D7, 0xE22B20D2, 0xE6EA3D65,
	0xEBA91BBC, 0xEF68060B, 0xD727BBB6, 0xD3E6A601, 0xDEA580D8,
	0xDA649D6F, 0xC423CD6A, 0xC0E2D0DD, 0xCDA1F604, 0xC960EBB3,
	0xBD3E8D7E, 0xB9FF90C9, 0xB4BCB610, 0xB07DABA7, 0xAE3AFBA2,
	0xAAFBE615, 0xA7B8C0CC, 0xA379DD7B, 0x9B3660C6, 0x9FF77D71,
	0x92B45BA8, 0x9675461F, 0x8832161A, 0x8CF30BAD, 0x81B02D74,
	0x857130C3, 0x5D8A9099, 0x594B8D2E, 0x5408ABF7, 0x50C9B640,
	0x4E8EE645, 0x4A4FFBF2, 0x470CDD2B, 0x43CDC09C, 0x7B827D21,
	0x7F436096, 0x7200464F, 0x76C15BF8, 0x68860BFD, 0x6C47164A,
	0x61043093, 0x65C52D24, 0x119B4BE9, 0x155A565E, 0x18197087,
	0x1CD86D30, 0x029F3D35, 0x065E2082, 0x0B1D065B, 0x0FDC1BEC,
	0x3793A651, 0x3352BBE6, 0x3E119D3F, 0x3AD08088, 0x2497D08D,
	0x2056CD3A, 0x2D15EBE3, 0x29D4F654, 0xC5A92679, 0xC1683BCE,
	0xCC2B1D17, 0xC8EA00A0, 0xD6AD50A5, 0xD26C4D12, 0xDF2F6BCB,
	0xDBEE767C, 0xE3A1CBC1, 0xE760D676, 0xEA23F0AF, 0xEEE2ED18,
	0xF0A5BD1D, 0xF464A0AA, 0xF9278673, 0xFDE69BC4, 0x89B8FD09,
	0x8D79E0BE, 0x803AC667, 0x84FBDBD0, 0x9ABC8BD5, 0x9E7D9662,
	0x933EB0BB, 0x97FFAD0C, 0xAFB010B1, 0xAB710D06, 0xA6322BDF,
	0xA2F33668, 0xBCB4666D, 0xB8757BDA, 0xB5365D03, 0xB1F740B4
};

/*
 * crcposix -- compute posix.2 compatible 32 bit CRC
 */
void
m_crcposix(crcp, bp, n)
ulong_t *crcp;
const uchar_t *bp;
size_t n;
{
	while (n-- > 0)
	    *crcp = (*crcp<<8) ^ crctab_posix[(uchar_t)((*crcp>>24)^*bp++)];
}
